File: shared/web-sdk/src/audio-pre-roll-ring.test.ts
import { describe, expect, it } from "vitest";
import { createAudioPreRollRing } from "./audio-pre-roll-ring.ts";

let counter = 0;
function frame(): ArrayBuffer {
  counter += 1;
  return new Uint8Array([counter & 0xff]).buffer;
}

function tags(buffers: readonly ArrayBuffer[]): number[] {
  return buffers.map((b) => new Uint8Array(b)[0] ?? -1);
}

describe("AudioPreRollRing FSM", () => {
  it("buffers rejected frames silently while idle and emits nothing", () => {
    const ring = createAudioPreRollRing({ preRollFrames: 3, hangoverFrames: 2 });
    expect(ring.push(frame(), false)).toEqual([]);
    expect(ring.push(frame(), false)).toEqual([]);
    expect(ring.push(frame(), false)).toEqual([]);
    expect(ring.push(frame(), false)).toEqual([]);
  });

  it("flushes the buffered pre-roll ahead of the first accepted frame", () => {
    const ring = createAudioPreRollRing({ preRollFrames: 3, hangoverFrames: 2 });
    const a = frame();
    const b = frame();
    const c = frame();
    const speech = frame();
    ring.push(a, false);
    ring.push(b, false);
    ring.push(c, false);
    const out = ring.push(speech, true);
    expect(tags(out)).toEqual([tags([a])[0], tags([b])[0], tags([c])[0], tags([speech])[0]]);
  });

  it("trims the pre-roll ring to the configured size (drops oldest)", () => {
    const ring = createAudioPreRollRing({ preRollFrames: 2, hangoverFrames: 0 });
    const dropped = frame();
    const b = frame();
    const c = frame();
    const speech = frame();
    ring.push(dropped, false);
    ring.push(b, false);
    ring.push(c, false);
    const out = ring.push(speech, true);
    expect(tags(out)).toEqual([tags([b])[0], tags([c])[0], tags([speech])[0]]);
    expect(tags(out)).not.toContain(tags([dropped])[0]);
  });

  it("emits each accepted frame as-is while active", () => {
    const ring = createAudioPreRollRing({ preRollFrames: 2, hangoverFrames: 2 });
    ring.push(frame(), true);
    const second = frame();
    const out = ring.push(second, true);
    expect(out.length).toBe(1);
    expect(tags(out)).toEqual([tags([second])[0]]);
  });

  it("emits hangover frames after speech ends, then falls silent", () => {
    const ring = createAudioPreRollRing({ preRollFrames: 3, hangoverFrames: 2 });
    ring.push(frame(), true);
    const tail1 = frame();
    const tail2 = frame();
    const afterTail = frame();
    expect(tags(ring.push(tail1, false))).toEqual([tags([tail1])[0]]);
    expect(tags(ring.push(tail2, false))).toEqual([tags([tail2])[0]]);
    expect(ring.push(afterTail, false)).toEqual([]); // counter exhausted, frame buffered
  });

  it("re-arms the hangover counter on accept-during-hangover", () => {
    const ring = createAudioPreRollRing({ preRollFrames: 3, hangoverFrames: 2 });
    ring.push(frame(), true);
    expect(ring.push(frame(), false).length).toBe(1); // hangover 1/2 used
    ring.push(frame(), true); // re-arm to 2
    expect(ring.push(frame(), false).length).toBe(1); // hangover 1/2 used
    expect(ring.push(frame(), false).length).toBe(1); // hangover 2/2 used
    expect(ring.push(frame(), false).length).toBe(0); // exhausted
  });

  it("reset clears the ring and returns to idle", () => {
    const ring = createAudioPreRollRing({ preRollFrames: 2, hangoverFrames: 2 });
    ring.push(frame(), false);
    ring.push(frame(), true);
    ring.reset();
    const afterReset = frame();
    const out = ring.push(afterReset, true);
    expect(out.length).toBe(1);
    expect(tags(out)).toEqual([tags([afterReset])[0]]);
  });

  it("with zero pre-roll and zero hangover behaves like a pass-through gate", () => {
    const ring = createAudioPreRollRing({ preRollFrames: 0, hangoverFrames: 0 });
    expect(ring.push(frame(), false)).toEqual([]);
    const speech = frame();
    expect(tags(ring.push(speech, true))).toEqual([tags([speech])[0]]);
    expect(ring.push(frame(), false)).toEqual([]);
  });

  it("works with a non-ArrayBuffer frame type (Float32Array)", () => {
    const ring = createAudioPreRollRing<Float32Array>({ preRollFrames: 2, hangoverFrames: 0 });
    const a = new Float32Array([1]);
    const b = new Float32Array([2]);
    const c = new Float32Array([3]);
    expect(ring.push(a, false)).toEqual([]); // buffered
    expect(ring.push(b, false)).toEqual([]); // buffered (ring full: [a,b])
    expect(ring.push(c, true)).toEqual([a, b, c]); // accept → flush pre-roll [a,b] + current c
  });
});

File: shared/mobile-sdk/src/commonTest/kotlin/io/sentient/mobilesdk/audio/AudioPreRollRingTest.kt
package io.sentient.mobilesdk.audio

import kotlin.test.Test
import kotlin.test.assertEquals

// ---------------------------------------------------------------------------
// AudioPreRollRingTest — port of web-sdk audio-pre-roll-ring.test.ts FSM tests.
//
// Pins: idle buffering, ring trim, onset flush, active passthrough,
// hangover countdown, hangover re-arm, reset, zero-config pass-through gate.
//
// Only cases that pin real buffer/FSM behavior are kept (test-lean doctrine).
// Vacuous wiring tests (types, constants, DI) are omitted.
// ---------------------------------------------------------------------------

class AudioPreRollRingTest {

    // -----------------------------------------------------------------------
    // Seed cases (from plan)
    // -----------------------------------------------------------------------

    @Test
    fun onset_flushes_preroll_then_frame() {
        val r = AudioPreRollRing<Int>(preRollFrames = 2, hangoverFrames = 1)
        assertEquals(emptyList(), r.push(1, accepted = false))
        assertEquals(emptyList(), r.push(2, accepted = false))
        assertEquals(emptyList(), r.push(3, accepted = false)) // ring trims to [2,3]
        assertEquals(listOf(2, 3, 4), r.push(4, accepted = true))
    }

    @Test
    fun hangover_emits_then_idles() {
        val r = AudioPreRollRing<Int>(preRollFrames = 0, hangoverFrames = 1)
        assertEquals(listOf(10), r.push(10, accepted = true))
        assertEquals(listOf(11), r.push(11, accepted = false)) // hangover emits 1
        assertEquals(emptyList(), r.push(12, accepted = false)) // idle now
    }

    // -----------------------------------------------------------------------
    // Idle buffering — rejected frames are buffered silently
    // -----------------------------------------------------------------------

    @Test
    fun buffers_rejected_frames_silently_while_idle() {
        val r = AudioPreRollRing<Int>(preRollFrames = 3, hangoverFrames = 2)
        assertEquals(emptyList(), r.push(1, accepted = false))
        assertEquals(emptyList(), r.push(2, accepted = false))
        assertEquals(emptyList(), r.push(3, accepted = false))
        assertEquals(emptyList(), r.push(4, accepted = false))
    }

    // -----------------------------------------------------------------------
    // Ring trim — oldest frames are dropped when ring overflows
    // -----------------------------------------------------------------------

    @Test
    fun ring_trims_to_preroll_frames_dropping_oldest() {
        val r = AudioPreRollRing<Int>(preRollFrames = 2, hangoverFrames = 0)
        // push 3 rejected: ring should hold only last 2
        r.push(10, accepted = false) // dropped
        r.push(20, accepted = false) // kept
        r.push(30, accepted = false) // kept
        val out = r.push(40, accepted = true)
        assertEquals(listOf(20, 30, 40), out)
    }

    // -----------------------------------------------------------------------
    // Flush — onset emits the pre-roll ahead of the triggering accepted frame
    // -----------------------------------------------------------------------

    @Test
    fun flushes_buffered_preroll_ahead_of_first_accepted_frame() {
        val r = AudioPreRollRing<Int>(preRollFrames = 3, hangoverFrames = 2)
        r.push(1, accepted = false)
        r.push(2, accepted = false)
        r.push(3, accepted = false)
        val out = r.push(4, accepted = true)
        assertEquals(listOf(1, 2, 3, 4), out)
    }

    // -----------------------------------------------------------------------
    // Active passthrough — each accepted frame is emitted as-is while active
    // -----------------------------------------------------------------------

    @Test
    fun emits_each_accepted_frame_as_is_while_active() {
        val r = AudioPreRollRing<Int>(preRollFrames = 2, hangoverFrames = 2)
        r.push(1, accepted = true) // onset
        val out = r.push(2, accepted = true)
        assertEquals(listOf(2), out)
    }

    // -----------------------------------------------------------------------
    // Hangover — emits tail frames, then falls silent after counter exhausted
    // -----------------------------------------------------------------------

    @Test
    fun emits_hangover_frames_after_speech_ends_then_falls_silent() {
        val r = AudioPreRollRing<Int>(preRollFrames = 3, hangoverFrames = 2)
        r.push(1, accepted = true) // onset
        assertEquals(listOf(2), r.push(2, accepted = false)) // hangover 1/2
        assertEquals(listOf(3), r.push(3, accepted = false)) // hangover 2/2
        assertEquals(emptyList(), r.push(4, accepted = false)) // exhausted, buffered
    }

    // -----------------------------------------------------------------------
    // Hangover re-arm — accept during hangover resets the counter
    // -----------------------------------------------------------------------

    @Test
    fun rearms_hangover_counter_on_accept_during_hangover() {
        val r = AudioPreRollRing<Int>(preRollFrames = 3, hangoverFrames = 2)
        r.push(1, accepted = true) // onset
        assertEquals(1, r.push(2, accepted = false).size) // hangover 1/2 used
        r.push(3, accepted = true) // re-arm hangover to 2
        assertEquals(1, r.push(4, accepted = false).size) // hangover 1/2 used
        assertEquals(1, r.push(5, accepted = false).size) // hangover 2/2 used
        assertEquals(0, r.push(6, accepted = false).size) // exhausted
    }

    // -----------------------------------------------------------------------
    // Reset — clears ring and returns to idle
    // -----------------------------------------------------------------------

    @Test
    fun reset_clears_ring_and_returns_to_idle() {
        val r = AudioPreRollRing<Int>(preRollFrames = 2, hangoverFrames = 2)
        r.push(1, accepted = false)
        r.push(2, accepted = true) // active
        r.reset()
        // After reset, onset should flush empty pre-roll (ring was cleared)
        val out = r.push(3, accepted = true)
        assertEquals(listOf(3), out)
    }

    // -----------------------------------------------------------------------
    // Zero-config — preRollFrames=0, hangoverFrames=0 is a pass-through gate
    // -----------------------------------------------------------------------

    @Test
    fun zero_preroll_zero_hangover_behaves_as_pass_through_gate() {
        val r = AudioPreRollRing<Int>(preRollFrames = 0, hangoverFrames = 0)
        assertEquals(emptyList(), r.push(1, accepted = false))
        assertEquals(listOf(2), r.push(2, accepted = true))
        assertEquals(emptyList(), r.push(3, accepted = false))
    }
}
